File: examples/go/multisig-2of3/MultiSig2of3_test.go
package contract

import (
	"testing"

	runar "github.com/icellan/runar/packages/runar-go"
)

// R-107 — `multisig-2of3` is the canonical checkMultiSig + array-literal
// example. The two array arguments are the canonical site where the
// `array_literal` ANF node is emitted, and `array_literal` is one of the four
// node kinds `spec/ir-format.md` did not document until R-098.
//
// MultiSig2of3.runar.go carries //go:build ignore, so the contract type is not
// constructible from this test binary and this suite covers the Rúnar frontend
// (parse → validate → typecheck) only.
//
// The reason is written at the top of the contract file: the `[N]T{...}`
// composite literal that three of the seven .runar.go parsers require does not
// convert to the `[]Sig` / `[]PubKey` the mock CheckMultiSig takes. The slice
// spelling was tried here and reverted — it builds as Go and the Go tier emits
// identical ANF, but the TypeScript, Zig and Ruby parsers reject it. Native
// business-logic tests for the ordered-pair semantics are what this exclusion
// costs; they were written, they passed, and they are not reachable while the
// type cannot be constructed.

func TestMultiSig2of3_Compile(t *testing.T) {
	if err := runar.CompileCheck("MultiSig2of3.runar.go"); err != nil {
		t.Fatalf("Rúnar compile check failed: %v", err)
	}
}

File: examples/rust/multisig-2of3/MultiSig2of3_test.rs
// R-107 — `multisig-2of3` is the canonical checkMultiSig + array-literal
// example and was tested in four of the nine formats (ts, sol, move, zig).
// This is the Rust half.
//
// `check_multi_sig([sig1, sig2], [self.pk1, self.pk2, self.pk3])` lowers to two
// `array_literal` ANF nodes — the canonical site where that node kind is
// emitted at all.
//
// The contract's `check_multi_sig` is a Rúnar intrinsic the compiler
// materialises into Script; it is not a method on the macro-generated struct,
// so `#[path] mod contract;` would not build. The frontend is the boundary the
// cross-compiler conformance cares about, and `compile_check` is how the peer
// suites exercise it.

#[test]
fn test_compile() {
    runar::compile_check(
        include_str!("MultiSig2of3.runar.rs"),
        "MultiSig2of3.runar.rs",
    )
    .unwrap();
}
